File: demos/BirdRecord.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Garden.Birds;

public enum Habitat { Garden, Woodland, Wetland }

public sealed record Sighting(string Species, int Count, Habitat Habitat, DateTimeOffset SeenAt)
{
    public string Label => $"{Count} {Species} in the {Habitat.ToString().ToLowerInvariant()}";
}

public static class BirdReport
{
    public static IReadOnlyDictionary<Habitat, int> Summarize(IEnumerable<Sighting> sightings) =>
        sightings.GroupBy(sighting => sighting.Habitat)
            .ToDictionary(group => group.Key, group => group.Sum(sighting => sighting.Count));

    public static void Main()
    {
        var sightings = new[] { new Sighting("Bullfinch", 2, Habitat.Garden, DateTimeOffset.UtcNow) };
        Console.WriteLine(string.Join(", ", Summarize(sightings).Select(pair => $"{pair.Key}: {pair.Value}")));
    }
}

File: demos/BirdRecord.java
import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public final class BirdRecord {
  enum Habitat { GARDEN, WOODLAND, WETLAND }

  record Sighting(String species, int count, Habitat habitat, Instant seenAt) {
    String label() {
      return "%d %s in %s".formatted(count, species, habitat.name().toLowerCase());
    }
  }

  static Map<Habitat, Integer> summarize(List<Sighting> sightings) {
    return sightings.stream().collect(Collectors.groupingBy(
        Sighting::habitat,
        Collectors.summingInt(Sighting::count)
    ));
  }

  public static void main(String[] args) {
    var sightings = List.of(new Sighting("Bullfinch", 2, Habitat.GARDEN, Instant.now()));
    summarize(sightings).forEach((habitat, count) -> System.out.println(habitat + ": " + count));
  }
}
